### PR TITLE
Add endpoint to retrieve aggregated environment metrics from the server

### LINKED ISSUE #5129: Add API endpoint to retrieve metrics
This tickets add the `/metrics` API endpoint used by the client to retrieve the metrics within a certain time interval.

Tasks:

- [x] Add the `/metrics` endpoint
- [x] Ensure that this endpoint has support to perform the retrieval-time aggregation for all MetricsTypes:
       - counter metric
       - non-counter metric
       - compile_rate metric
- [x] The size of the interval passed to the `/metrics` endpoint should be at least: `nb_datapoints * 1 minute` (1 minute is the data collection interval)

Make sure the following corner cases are tested:

- [x] No metrics data is available for a certain sub-interval
- [x] Partial metrics data is available for a certain sub-interval
- [x] The timestamp of a datapoint overlaps with the boundary of a sub-interval

### PR DESCRIPTION
**Pull request by arnaudsjs**
# Description

Add endpoint to retrieve aggregated environment metrics from the server

closes #5129

# Self Check:

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [x] End user documentation is included or an issue is created for end-user documentation

### CODE REVIEW

File: src/inmanta/data/model.py [hunk 1]
```diff
@@ -702,3 +702,23 @@ class Source(BaseModel):
...
+
+
+class EnvironmentMetricsResult(BaseModel):
+    """
+    A container for metrics as returned by the /metrics endpoint.
+
+    :param start: The starting of the requested aggregation interval.
+    :param end: The end of the requested aggregation interval.
+    :param timestamps: The timestamps that belongs to the aggregated metrics present in the `metrics` dictionary.
+    :param metrics: A dictionary that maps the name of a metric a list of aggregated datapoints. For metrics that are not
```
Comment (FloLey) [inline src/inmanta/data/model.py:714]:
```suggestion
    :param metrics: A dictionary that maps the name of a metric to a list of aggregated datapoints. For metrics that are not
```

File: tests/server/test_environment_metrics.py [hunk 2]
```diff
@@ -930,3 +933,364 @@ async def add_compiles(environment: uuid.UUID, wait_times: abc.Sequence[float]):
...
+    assert result.code == 200, result.result
+    assert datetime.fromisoformat(result.result["data"]["start"]) == get_as_naive_datetime(start_interval)
+    assert datetime.fromisoformat(result.result["data"]["end"]) == get_as_naive_datetime(start_interval_plus_1h)
+    expected_timestamps = [get_as_naive_datetime(start_interval + timedelta(minutes=(i + 1) * 6)) for i in range(nb_datapoints)]
+    assert [datetime.fromisoformat(timestamp) for timestamp in result.result["data"]["timestamps"]] == expected_timestamps
+    assert len(result.result["data"]["metrics"]) == 2
+    assert result.result["data"]["metrics"]["gauge_metric1"] == [sum(i for _ in range(6)) / 6 + 0.5 for i in range(10)]
+    assert result.result["data"]["metrics"]["timer_metric1"] == [(sum(i for _ in range(6)) + 1.5) / (2 * 6) for i in range(10)]
+
+    # Verify behavior when no data is available a time window
```
Comment (FloLey) [inline tests/server/test_environment_metrics.py:1159]:
```suggestion
    # Verify behavior when no data is available in the time window
```

File: tests/server/test_environment_metrics.py [hunk 3]
```diff
@@ -930,3 +933,364 @@ async def add_compiles(environment: uuid.UUID, wait_times: abc.Sequence[float]):
...
+    )
+    assert result.code == 200, result.result
+    assert datetime.fromisoformat(result.result["data"]["start"]) == get_as_naive_datetime(start_interval_min_6_min)
+    assert datetime.fromisoformat(result.result["data"]["end"]) == get_as_naive_datetime(start_interval_plus_6_min)
+    expected_timestamps = [get_as_naive_datetime(start_interval), get_as_naive_datetime(start_interval_plus_6_min)]
+    assert [datetime.fromisoformat(timestamp) for timestamp in result.result["data"]["timestamps"]] == expected_timestamps
+    assert len(result.result["data"]["metrics"]) == 1
+    assert result.result["data"]["metrics"]["gauge_metric1"] == [None, 0.5]
+
+    # Verify behavior when partial data is available a time window
```
Comment (FloLey) [inline tests/server/test_environment_metrics.py:1178]:
```suggestion
    # Verify behavior when partial data is available in the time window
```

**Review by FloLey — APPROVED** [review]

File: tests/utils.py [hunk 4]
```diff
@@ -653,3 +654,12 @@ async def get_data(postgresql_client):
     )
     assert post_ra_one_set == post_ra_two_set
     assert post_ra_one_set
+
+
+def get_as_naive_datetime(timestamp: datetime) -> datetime:
+    """
+    Convert the give timestamp, which is timezone aware, into a naive timestamp object.
```
Comment (sanderr) [inline tests/utils.py:661]:
```suggestion
    Convert the give timestamp, which is timezone aware, into a naive timestamp object in UTC.
```

File: tests/utils.py [hunk 5]
```diff
@@ -653,3 +654,12 @@ async def get_data(postgresql_client):
     )
     assert post_ra_one_set == post_ra_two_set
     assert post_ra_one_set
+
+
+def get_as_naive_datetime(timestamp: datetime) -> datetime:
```
Comment (sanderr) [inline tests/utils.py:659]: Do you use this anywhere? I don't see it.

**Review by sanderr — CHANGES_REQUESTED** [review]
Overall this looks good to me. Some minor comments.

File: src/inmanta/protocol/rest/__init__.py [hunk 6]
```diff
@@ -190,9 +190,23 @@ async def process(self) -> None:
...
+                            and self._properties.operation == "GET"
+                            and typing_inspect.is_generic_type(arg_type)
+                            and issubclass(typing_inspect.get_origin(arg_type), list)
+                            and not isinstance(self._message[arg], list)
+                            and len(typing_inspect.get_args(arg_type, evaluate=True)) == 1
+                            and isinstance(self._message[arg], typing_inspect.get_args(arg_type)[0])
+                        ):
+                            # If a GET endpoint has a parameter of type list that is encoded as a URL query parameter and the
+                            # specific request provides a list with one element, urllib doesn't parse it as a list.
+                            # Map it here explicitly to a list.
```
[the remark this answers is not in the document]
Comment (arnaudsjs) [reply src/inmanta/protocol/rest/__init__.py:206]: The query argument of type List is only an issue when the method type if `GET`. Up to now, all endpoint with a query argument of type list didn't have the `GET` method type.

File: tests/utils.py (hunk 5 above)
Comment (arnaudsjs) [reply tests/utils.py:659]: It's used in `tests/server/test_environment_metrics.py`

Comment (sanderr) [reply tests/utils.py:659]: For some reason I'd overlooked that it was in `tests/utils.py`, not `src/inmanta/util.py` so I was looking in all the wrong places.

File: src/inmanta/server/services/environment_metrics_service.py [hunk 7]
```diff
@@ -262,6 +267,132 @@ def create_metrics_timer(metric_values_timer: Sequence[MetricValueTimer], timest
...
+            table_name=EnvironmentMetricsTimer.table_name(),
+            aggregation_function="(sum(value)::float)/(sum(count)::float)",
+            metrics_list="$5",
+        )
+        query_for_compiler_rate = _get_sub_query(
+            metric="'orchestrator.compile_rate'",
+            group_by=f"'{DEFAULT_GROUPED_BY}'",
+            table_name=EnvironmentMetricsTimer.table_name(),
+            aggregation_function=(
+                "(sum(count)::float) / ((EXTRACT(epoch FROM ($3::timestamp - $2::timestamp)))::float / 3600)::float"
```
[the remark this answers is not in the document]
Comment (arnaudsjs) [reply src/inmanta/server/services/environment_metrics_service.py:351]: Oh. Nice catch!

File: src/inmanta/server/services/environment_metrics_service.py [hunk 8]
```diff
@@ -262,6 +267,132 @@ def create_metrics_timer(metric_values_timer: Sequence[MetricValueTimer], timest
...
+            group_by="grouped_by",
+            table_name=EnvironmentMetricsGauge.table_name(),
+            aggregation_function="(sum(count)::float)/(count(*)::float)",
+            metrics_list="$5",
+        )
+        query_on_timer_table = _get_sub_query(
+            metric="metric_name",
+            group_by="grouped_by",
+            table_name=EnvironmentMetricsTimer.table_name(),
+            aggregation_function="(sum(value)::float)/(sum(count)::float)",
```
[the remark this answers is not in the document]
Comment (arnaudsjs) [reply src/inmanta/server/services/environment_metrics_service.py:343]: Fixed.

**Review by sanderr — APPROVED** [review]

**Pull request closed without merging**